dds_web/security/project_keys.py

```diff
@@ -0,0 +1,144 @@
+""" Code for generating project related keys """
+import os
+
+import argon2
+import cryptography
+import flask
+import gc
+
+import dds_web
+from dds_web.errors import KeyNotFoundError
+
+
+def derive_key(user, password):
+    """Derive a KEK from users password."""
+    # Salt used for key derivation
+    if user.kd_salt is None:
+        exception = Exception("Access to project data is not properly setup for the user!")
+        flask.current_app.logger.exception(exception)
+        raise exception
+
+    # Derive the key
+    derived_key = argon2.low_level.hash_secret_raw(
+        secret=password.encode(),
+        salt=user.kd_salt,
+        time_cost=2,
+        # memory_cost=4194304, # This makes it crash locally
+        memory_cost=1000,  # Just changed to something so that it wouldn't crash
+        parallelism=8,
+        hash_len=32,
+        type=argon2.Type.ID,
+    )
+
+    # Verify length
+    if len(derived_key) != 32:
+        exception = Exception("Derived key is not 256 bits long!")
+        flask.current_app.logger.exception(exception)
+        raise exception
+
+    return derived_key
+
+
+def manage_project_key_among_users(existing_user, current_user, project_key):
+    return encrypt_project_key(existing_user, obtain_project_private_key(current_user, project_key))
+
+
+def obtain_project_private_key(user, project_key):
+    """Decrypt the project private key with the user specific KEK."""
+    # password = dds_web.cache.get(user.username)
+    password = "password"
```

Comment:
This one warrants a `TODO: please fix` don't you think?

Reply:
Sure 👍🏻